File: 04-trees-and-graphs/02_lowest_common_ancestor/LowestCommonAncestor.java
/**
 * Problem: Lowest Common Ancestor of a Binary Tree
 * LeetCode: #236 (Medium)
 * 
 * Time Complexity: O(N)
 * Space Complexity: O(H)
 */

public class LowestCommonAncestor {
    public static class TreeNode {
        int val;
        TreeNode left, right;
        TreeNode(int val) { this.val = val; }
    }

    public static TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null || root == p || root == q) return root;

        TreeNode left = lowestCommonAncestor(root.left, p, q);
        TreeNode right = lowestCommonAncestor(root.right, p, q);

        if (left != null && right != null) return root;
        return (left != null) ? left : right;
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(3);
        root.left = new TreeNode(5);
        root.right = new TreeNode(1);
        root.left.left = new TreeNode(6);
        root.left.right = new TreeNode(2);

        TreeNode p = root.left; // 5
        TreeNode q = root.left.right; // 2

        TreeNode lca = lowestCommonAncestor(root, p, q);
        System.out.println("[Java] LCA Test");
        System.out.println("LCA of " + p.val + " and " + q.val + " is " + lca.val);
    }
}

File: 04-trees-and-graphs/02_lowest_common_ancestor/lowest_common_ancestor.c
/**
 * Problem: Lowest Common Ancestor of a Binary Tree
 * LeetCode: #236 (Medium)
 * 
 * Time Complexity: O(N)
 * Space Complexity: O(H) recursion stack
 */

#include <stdio.h>
#include <stdlib.h>

struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

struct TreeNode* createNode(int val) {
    struct TreeNode* node = (struct TreeNode*)malloc(sizeof(struct TreeNode));
    node->val = val;
    node->left = NULL;
    node->right = NULL;
    return node;
}

struct TreeNode* lowestCommonAncestor(struct TreeNode* root, struct TreeNode* p, struct TreeNode* q) {
    if (!root || root == p || root == q) return root;

    struct TreeNode* left = lowestCommonAncestor(root->left, p, q);
    struct TreeNode* right = lowestCommonAncestor(root->right, p, q);

    if (left && right) return root;
    return left ? left : right;
}

int main(void) {
    struct TreeNode* root = createNode(3);
    root->left = createNode(5);
    root->right = createNode(1);
    root->left->left = createNode(6);
    root->left->right = createNode(2);
    root->right->left = createNode(0);
    root->right->right = createNode(8);

    struct TreeNode* p = root->left;        // 5
    struct TreeNode* q = root->left->right; // 2

    struct TreeNode* lca = lowestCommonAncestor(root, p, q);
    printf("[C] Lowest Common Ancestor Test\n");
    printf("LCA of %d and %d is: %d (Expected: 5)\n", p->val, q->val, lca ? lca->val : -1);
    return 0;
}
